### ISSUE TITLE
[Android] Joplin not syncing notes properly - gives up after syncing less than half of the notes

### ISSUE THREAD

**Issue by N0tallion6** [laurent22/joplin#8730]
I've been using Joplin as my primary note-taking app for a while now. I have my notes synced between my desktop, phone and my laptop using Syncthing as my primary method of synchronisation. I use the 'File System' option to sync my notes.

Ever since 2.11 was released I have not for the life of me been able to get my notebooks to sync to my phone. At least not completely.

At first when I initially updated to the latest Joplin version on Android, (Shortly after updating it to the latest version on my desktop and laptop) some notes were showing as still encrypted and would not de-crypt no matter how many times I pressed the 'Synchronise' button. This was also an issue before I updated on Android, but I figured that the update would solve it and I could just get on with my day. Unfortunately it was not that simple.
After a few fruitless attempts, I opted to uninstall the app and then re-install, importing my entire notebook from scratch.

This is where the problem got worse - after re-installing the app, pointing it to my notes library and putting in my decryption password, (I waited about 3 minutes before it prompted me) I waited a long time for it to start importing my notes. Even after waiting upwards of half an hour for the sync to start, it only synced a handful of my notes and then just gave up. It didn't even import any of my folder structures, just a handful of unrelated notes. All I have in the hamburger menu is the 'All Notes' option and a single folder labelled 'Notebooks', but I cannot open it, presumably because Joplin still thinks it's empty.

I tried resetting the app multiple times, still get the same result. I even tried deleting and re-syncing my entire notes library in case something on there had corrupted at some point. Again, still the same result.

The weirdest thing about this problem is that there are no errors at all, Joplin just seems to stop without warning. Last time I tried 
```
Fetched Items: 126/150
Completed 21/08/2023 16:03
(826s)

```
As you can see from the above, it just gives up at an odd and seemingly random number. Most it's ever reached in my numerous attempts is somewhere shy of 450. I know for a fact I have a LOT more notes than that - the Encryption dialog on my PC in the settings tells me that I have a total of 3961 items, while the Android client seems to think I only have 449...)

I'm aware that background sync isn't supported yet and I've tried syncing a multiple times whilst leaving my phone open next to me, tapping the screen occasionally to make sure it doesn't go to sleep, but again, same result.

Most I've ever been able to sync is short of 450 notes and even then it stops at an awkward number. It typically doesn't manage to get beyond ~150 notes) 

I even tried disabling battery optimisation in case that was doing anything, but again, same result.

My notes are still syncing fine between Joplin on my Desktop PC and my laptop, running Manjaro & Arch Linux respectively. 

Environment:

Joplin version: 2.11.31
Platform: Android 12
OS specifics: Motorola Moto G Pro Stylus

Steps to reproduce:

1.  Install Joplin
2.  Synchronise notes library via Syncthing
3.  Open Joplin, set notes location to where my notes are located on my phone.
4.  Press Sync button
5.  Wait for Joplin to prompt for Master password, put in Master password
6.  Wait for notes to sync (but Joplin gets to around 150 notes and just gives up)

Expected result:
Joplin notes eventually sync:

What actually happened:
Joplin proceeded to sync ~150 notes before giving up entirely

Any ideas on how I can resolve this? (Assuming that it's not a bug with the app itself)

Thanks

**Comment by wh201906** [comment]
Could you please provide the log?
https://joplinapp.org/debugging/#mobile-application

**Comment by N0tallion6** [comment]
[syncReport-1693176691492.txt](https://github.com/laurent22/joplin/files/12490663/syncReport-1693176691492.txt)

Sorry for the late reply @wh201906 , here's the log as requested

**Comment by wh201906** [comment]
Sorry but I can't find errors related to this issue in the log.
Maybe you could ask in the forum 
https://discourse.joplinapp.org/

**Comment by diegofps** [comment]
Hi @N0tallion6. I've been facing a similar problem since February but managed to fix it today. 

In my case, it hung on 135/139 items (I had around 310). The synchronize icon would spin forever and clicking "Cancel" would make no difference. It would continue to spin with the word Cancelling below it. What I did was close all my Joplin instances and erase the lock files I had in my synchronization folder (Dropbox in my case):

1. Close all your Dropbox instances. For the Android app, I opened the "App info" page in the system and clicked "Force stop".
2. Open the folder where Joplin saves your encrypted data and rename the folders "locks" and "temp" to "locks2" and "temp2". they were in "\<Dropbox>/App/Joplin". I had 4 locks in the lock folder even though I had no app running.
3. Create new empty folders with the names locks and temp. I tried to synchronize without them but Joplin would complain that he couldn't create some files. 
4. Attempt to synchronize your Android app. If it works, remove the folders locks2 and temp2.

* Make sure there are no instances executing when you do this, or you may damage your notes. You may want to backup the entire Joplin folder beforehand.

I am not a Joplin contributor, but here are a few ideas that come to my mind that may help to improve this process:

1. The locks could have a timestamp indicating when they were created. If they exist for at most a few minutes during the synchronization phase, maybe it makes sense to automatically ignore and delete older locks (more than 60 minutes old?). Similarly, if the synchronization process is taking more than 10 minutes, we could refresh the timestamp in the lock every time it gets older than 10 minutes.
2. The app should be able to synchronize fine when the locks and temp folder do not exist. It doesn't seem like a big deal to check if they exist during the start of the synchronization process and create them if necessary.

**Comment by N0tallion6** [comment]
Hi @diegofps , Thanks for your reply. Unfortunately, I followed the steps from your comment and that doesn't appear to have helped at all - Still takes forever to sync before giving up at a strange number. No particular errors in the sidebar.

I've Done some further digging and looked at the error log in the settings (Admittedly I should've paid a bit more attention to it - apologies to everyone trying to help me there!) and one of the errors is the following:

`09-08T15:13:25: Synchroniser: "Could not refresh lock - cancelling sync. Error was:",
"Error: Lock has expired`
Also sometimes it reports that it "cannot acquire the sync lock".

Another thing that's interesting, or perhaps I'm out of the loop, when browsing for my notes directory the path is displayed in my joplin settings as `content://con.android.externalstorage.documents/tree/primary:Documents/joplin`
I find this a little weird, because previously (before version 2.11it was reported as `/storage/0/emulated/Documents/joplin` - Is this part of the changes made with the latest version or is this potentially an indicator of the problem?

I since uninstalled Joplin, deleted and re-copied my notes to my phone, then re-installed Joplin, (latest version 2.11.31) double-checking that I've given Joplin all of the permissions it needs. Alas, still the same problem when I sync.

**Comment by N0tallion6** [comment]
Any ideas? Joplin on Android is practically unusable for me as a result of this issue...

**Comment by diegofps** [comment]
I still think you have a lock issue, or at least this is what your log tells me in the following two errors:

```
08-27T23:48:26,20,"Synchronizer: ""Could not refresh lock - cancelling sync. Error was:", "Error: Lock has expired
Code: lockExpired
Error: Lock has expired
    at construct (native)
    ...
```

and

```
>08-16T13:20:54,10,"Synchronizer: ""Error: Cannot acquire sync lock: either the lock could be written but not read back. Or it was expired before it was read again.
Error: Cannot acquire sync lock: either the lock could be written but not read back. Or it was expired before it was read again.
    at ?anon_0_ (address at index.android.bundle:1:2058253)
    ...
```

Using the **dev** branch as a reference, the first error is emitted by **packages/lib/Synchronizer.ts:520**, and the second error is emitted by **packages/lib/services/synchronizer/LockHandler.ts:385**. I believe they are both happening as activeLock (**packages/lib/services/synchronizer/LockHandler.ts:65**) fails to find a suitable Lock.

This gives me the following ideas:

1. Check the date and time of all your systems, a wrong date may be causing expired behavior.
2. If it is a Lock issue, clearing the Locks in the synchronization folder and reinstalling the Android Jopling should be enough to recreate the Lock correctly and make it work (creating a backup first).
3. If there is an error during the synchronization process that prevents Joplin from releasing the Lock after the first sync, it would make the following logs contain multiple Lock errors but not the root cause of the Lock issue. In this case, I would try the following: (1) remove Android Joplin; (2) clear the Locks in the cloud folder; (3) Install Android Joplin and let it sync; (4) export the log after the first or second failed sync, this would contain the original error corrupting the Lock.

Of course, there may be something entirely different happening here and... 

If you are a developer, you could also try to build the project yourself, add more log messages, and try to locate more info about the error when you execute it on an emulator. This is actually valuable as you learn a lot about working on a big project, react-native, typescript, and other cool things this project uses.

**Comment by N0tallion6** [comment]
@diegofps 
Date and time on both my phone and my computer(s) are exactly the same as far as I can tell...

I've Uninstalled/Re-installed Joplin on my phone multiple times now, but still get the same result, I hoped updating to version 2.12.3 would help in some way but no dice.
I re-installed again, Joplin doesn't even prompt me for the decryption key now either - I have to go into the settings and set it manually. Not sure if that has any bearing on the main issue, but just an observation in case.

I'm not even sure what's going on with the locks anymore - I used the splitscreen function on my Android phone to view the locks folder while Joplin was syncing. A lock file was created (the `locks` directory was empty prior to this) and seemed to persist for a while with the timestamp updating every few minutes, but when the lock error occurs and the sync cancels itself , Joplin seems to delete the lock file, so in most cases there isn't even a lock file for me to manually delete afterwards!

As suggested, I re-installed the app, attempted to re-sync and took the errors from the log of the 2nd attempt:
```
10-03T17:00:50,30,"Synchronizer: ""Sync: cancelling""
10-03T17:00:50,20,"Synchronizer: ""Could not refresh lock - cancelling sync. Error was:", "Error: Lock has expired
Code: lockExpired
Error: Lock has expired
    at construct (native)
    at apply (native)
    at _construct (address at index.android.bundle:1:1864056)
    at Wrapper (address at index.android.bundle:1:1863722)
    at construct (native)
    at _createSuperInternal (address at index.android.bundle:1:1862815)
    at call (native)
    at JoplinError (address at index.android.bundle:1:1863176)
    at ?anon_0_ (address at index.android.bundle:1:2060445)
    at next (native)
    at fulfilled (address at index.android.bundle:1:2056091)
    at tryCallOne (/root/react-native/ReactAndroid/hermes-engine/.cxx/MinSizeRel/164673tg/arm64-v8a/lib/InternalBytecode/InternalBytecode.js:53:16)
    at anonymous (/root/react-native/ReactAndroid/hermes-engine/.cxx/MinSizeRel/164673tg/arm64-v8a/lib/InternalBytecode/InternalBytecode.js:139:27)
    at apply (native)
    at anonymous (address at index.android.bundle:1:214542)
    at _callTimer (address at index.android.bundle:1:213491)
/
    at _callReactNativeMicrotasksPass (address at index.android.bundle:1:213655)
    at callReactNativeMicrotasks (address at index.android.bundle:1:215630)
    at __callReactNativeMicrotasks (address at index.android.bundle:1:116014)
    at anonymous (address at index.android.bundle:1:115107)
    at __guard (address at index.android.bundle:1:115855)
    at flushedQueue (address at index.android.bundle:1:115018)
    at invokeCallbackAndReturnFlushedQueue (address at index.android.bundle:1:114961)""
```


It looks like a lock issue like you say judging by the errors, though I can't think what might be preventing the locks from being refreshed. As mentioned I never had this problem at all prior to AES-256 being implemented. 

I'll attach the full log here as well just in case that helps at all...
[syncReport-1696348931811.txt](https://github.com/laurent22/joplin/files/12796774/syncReport-1696348931811.txt)

Let me know if there's anything else you want me to try!

> If you are a developer, you could also try to build the project yourself, add more log messages, and try to locate more info about the error when you execute it on an emulator. This is actually valuable as you learn a lot about working on a big project, react-native, typescript, and other cool things this project uses.

Unfortunately I'm not really a developer. My coding knowledge doesn't really go beyond some basic python and shell-scripting under Linux.

**Comment by N0tallion6** [comment]
Got a notification on my phone today telling me that a new version of Joplin had been released on F-Droid (version `2.13.5`). Thought I'd download it and give this another try.
This time Joplin was able to sync a more than the previous ~447 notes, getting to around 1000, so that's a step in the right direction, but it's still not pulling through any folders and still giving me the `Could not refresh lock - cancelling sync` error before pulling everything through.

Tried resetting the app (again) and re-doing from scratch, but no dice, still throws the same error and stops at a seemingly odd number. (only got to just over 600 notes) Checked the lock folder and still no locks in there to even delete.

What's interesting however is that there is a certain group of notes, older than a couple of years that always seem to pull through regardless - might be a clue as to what's wrong?

EDIT: Any further attempts since this post seem to now be just as bad as when I initially raised this issue.
